services/inc/enumclass.h

```diff
@@ -22,12 +22,225 @@
 
 namespace particle {
 
-template <typename E>
-constexpr auto to_underlying(E e) noexcept
-{
+template <typename T, typename Enable = void>
+class BitMaskFlags;
+
+template<typename T>
+struct EnableBitwise {
+    static const bool enable = false;
+};
+
+#define ENABLE_ENUM_CLASS_BITWISE(T) \
+template<>                           \
+struct EnableBitwise<T> {            \
+    static const bool enable = true; \
+}
+
+template<typename T>
+class BitMaskFlags<T, typename std::enable_if_t<std::is_enum<T>::value>> {
+public:
+    typedef typename std::underlying_type<T>::type ValueType;
+    
+    BitMaskFlags();
+    BitMaskFlags(const T& value);
+    BitMaskFlags(const BitMaskFlags<T>& flags);
+
+    BitMaskFlags<T> operator&(const BitMaskFlags<T>& flags) const;
+    BitMaskFlags<T>& operator&=(const BitMaskFlags<T>& flags);
+
+    BitMaskFlags<T> operator|(const BitMaskFlags<T>& flags) const;
+    BitMaskFlags<T>& operator|=(const BitMaskFlags<T>& flags);
+
+    BitMaskFlags<T> operator^(const BitMaskFlags<T>& flags) const;
+    BitMaskFlags<T>& operator^=(const BitMaskFlags<T>& flags);
+
+    BitMaskFlags<T> operator~() const;
+
+    BitMaskFlags<T> operator<<(size_t n) const;
+    BitMaskFlags<T>& operator<<=(size_t n);
+    BitMaskFlags<T> operator>>(size_t n) const;
+    BitMaskFlags<T>& operator>>=(size_t n);
+
+    typename std::underlying_type<T>::type value() const;
+
+    operator bool() const;
+    operator ValueType() const;
```

Comment:
We should remove both these operators, as otherwise we lose the type-safety by enabling implicit casting to bool and `ValueType`. Separate methods like `bool isSet()` and `value()` are a better choice even if require a bit more typing when using this class.

Reply:
I agree that we remove the `ValueType` operator. While the `bool` operator is very useful when we are doing something like:
```
if (enumFlags & ENUM_VALUE) {
}
```

Reply:
That's won't call the `bool` operator though. It uses `ValueType` operator.

Reply:
What if I remove the `ValueType` operator? Can we still keep the `bool` operator?

Reply:
I think we will still lose type-safety, because int <-> bool implicit conversions will still happen.